lib/decimal.ex

```diff
@@ -261,13 +261,35 @@ defmodule Decimal do
   def inf?(%Decimal{coef: :inf}), do: true
   def inf?(%Decimal{}), do: false
 
-  @doc """
-  Returns `true` if argument is a decimal number, otherwise `false`.
-  """
+  @doc false
   @spec decimal?(any) :: boolean
+  @deprecated "Use Decimal.is_decimal/1 macro instead"
   def decimal?(%Decimal{}), do: true
   def decimal?(_), do: false
 
+  @doc """
+  Returns `true` if argument is a decimal number, otherwise `false`.
+
+  ## Examples
+
+      iex> Decimal.is_decimal(Decimal.new(42))
+      true
+
+      iex> Decimal.is_decimal(42)
+      false
+
+  """
+  doc_since("1.9.0")
+  # TODO: make it similar to Kernel.is_struct when we require Elixir v1.10
+  defmacro is_decimal(term) do
+    quote do
+      case unquote(term) do
+        %Decimal{} -> true
+        _ -> false
+      end
+    end
+  end
```

Comment:
We should check `__CALLER__.context == :guard` and use the new map guard functions.

Reply:
I believe the new guards require more recent OTP versions. We could ship this to start people on migrating, add conditional logic to make this macro more useful on recent OTP, or wait with this change until decimal v2.0 where I assume we would require more recent Elixir. I think conditional logic or waiting is probably better, wdyt?

Reply:
@ericmj ping

Reply:
It's a bit confusing that it is named `is_` but doesn't work in guards. We can do conditional compilation against 1.10 and use the new is_struct / is_map_key functions. WDYT?

Reply:
Would also need to document that it only works as a guard on elixir 1.10.